hera_qm/xrfi.py

```diff
@@ -222,6 +224,223 @@ def robust_divide(num, den):
+    flag_method: str
+        Method used for flagging auto-correlations. Current options are 'dpss_flagger' and 'channel_diff_flagger'
+    nsig: float, default=10
+        The number of sigma in the metric above which to flag pixels.
+    int_count: float, default=None
+        Number of samples per integration in correlator (dt * dnu)
+    antenna_class: AntennaClassification, default=None
+        Optional AntennaClassification object. If provided, the flagging method chosen will skip antennas marked "bad"
+    wgts: dictionary or DataContainer: default=None
+        Dictionary or DataContainer containing data weights. If wgts="None", then weights will be set to ones
+    array_wgts_threshold: float, default=0.2
+        The fraction of flags required to trigger a broadcast across all auto-correlations for
+        a given (time, frequency) pixel in the combined weights array.
+    flag_kwargs: dict
+        Additional keyword arguments for the function chosen
+
+    Returns:
+    -------
+    wgts: dictionary or DataContainer
+        Dictionary or DataContainer containing weights for each antenna with values of either zero or one with
+        zero identifying channels found to have RFI
+    array_weights: np.ndarray
+        Contains weights for the entire array averaged over all antennas. If antenna_class is provide, only antennas
+        labeled 'good' or 'suspect' are included in the average    
+    """
+    from hera_cal.utils import split_bl
+    
+    # Check to see if flagging method is valid
+    if flag_method not in auto_flaggers:
+        raise ValueError("Flagging method must be a valid flagging function.")
+
+    alg_func = auto_flaggers[flag_method]
+
+    # If wgts is None, set weights to ones
+    wgts = {} if wgts is None else deepcopy(wgts)
+    rfi_wgts = {}
+
+    # Infer int_count from data's metadata if not provided
+    if int_count is None:
+        int_time = 24 * 3600 * np.median(np.diff(data.times))
+        chan_res = np.median(np.diff(data.freqs))
+        int_count = int(int_time * chan_res)
+
+    array_weights = []
+
+    # Iterate through data
+    for k, v in data.items():
+        ant1, ant2 = split_bl(k)
+        # Only flag auto-correlations
+        if ant1 == ant2:
+            # If antenna is labeled "bad", skip and set weights to zeros
+            if antenna_class is not None and antenna_class.is_bad(ant1):
+                rfi_wgts[ant1] = np.zeros_like(v)
+                continue
+
+            # Estimate the noise standard deviation from data
+            noise = np.abs(v) / np.sqrt(int_count / 2)
+
+            # Run flagging algorithm
+            rfi_wgts[ant1] = alg_func(v, noise=noise, nsig=nsig, wgts=wgts.get(ant1, np.ones_like(v)), **flag_kwargs)
+            array_weights.append((1 - rfi_wgts[ant1]))
+
+    # Combine into array-wide weights
+    array_weights = np.mean(array_weights, axis=0)
+    array_weights = np.where(array_weights < array_wgts_threshold, 1, 0)
```

Comment:
If you already know that an antenna is bad, should it count against you here? Like, should the threshold be over all antennas or over the not-already-known-to-be-bad antennas?